docs/proposals/ring-multikey.md

```diff
@@ -0,0 +1,148 @@
+      "addr": "10.0.0.1:9095",
+      "timestamp": 1660760278,
+      "tokens": [
+        1,
+        2
+      ],
+      "zone": "us-west-2b",
+      "registered_timestamp": 1660708390
+    },
+    "ingester-1": ...
+  }
+}
+```
+
+Proposal:
+```
+Key: ingester/ring/ingester-0
+Value:
+{
+    "addr": "10.0.0.1:9095",
+    "timestamp": 1660760278,
+    "tokens": [
+        1,
+        15
+    ],
+    "zone": "us-west-2b",
+    "registered_timestamp": 1660708390
+}
+
+Key: ingester/ring/ingester-1
+Value:
+{
+    "addr": "10.0.0.2:9095",
+    "timestamp": 1660760378,
+    "tokens": [
+        5,
+        28
+    ],
+    "zone": "us-west-2b",
+    "registered_timestamp": 1660708572
+}
+```
+
+The proposal is to create an interface called MultiKey. The interface allows KV store to request the codec to split and
+join the values is separated keys.
+
+```
+type MultiKey interface {
+    SplitById() map[string]interface{}
+
+    JoinIds(map[string]interface{}) Multikey
+
+    GetChildFactory() proto.Message
+
+    FindDifference(MultiKey) (Multikey, []string, error)
+}
+```
+
+* SplitById - responsible to split the codec in multiple keys and interface.
+* JoinIds - responsible to receive multiple keys and interface creating the codec objec
+* GetChildFactory - Allow the kv store to know how to serialize and deserialize the interface returned by “SplitById”.
+  The interface returned by SplitById need to be a proto.Message
+* FindDifference - optimization used to know what need to be updated or deleted from a codec. This avoids updating all keys every
+  time the coded change. First parameter returns a subset of the Multikey to be updated. Second is a list of keys to
+  be deleted.
+
+The codec implementation will change to support multiple keys. Currently, the codec interface for KV store supports
+only Encode and Decode. New methods will be added which would be used only by the KV stores implementing the multi
+key functionality.
+
+```
+type Codec interface {
+    //Existen
+    Decode([]byte) (interface{}, error)
+    Encode(interface{}) ([]byte, error)
+    CodecID() string
+
+    //Proposed
+    DecodeMultiKey(map[string][]byte) (interface{}, error)
+    EncodeMultiKey(interface{}) (map[string][]byte, error)
+}
+```
+
+* DecodeMultiKey - called by KV store to decode data downloaded. This function will use the JoinIds method.
+* EncodeMultiKey - called by KV store to encode data to be saved. This function will use the SplitById method.
+
+The new KV store will know the data being saved is a reference for multikey. It will use the FindDifference
+to know which keys need to be updated. The codec implementation for the new methods will use the JoinIds and SplitById
+to know how to separate the codec in multiple keys. The DecodeMultiKey will also use GetChildFactory to know how to
+decode the data stored in the kv store.
+
+Example of CAS being used with multikey design:
+![Sequence diagram](/images/proposals/ring-multikey-sequence.png)
```

Comment:
The ingesters generate hash tokens locally, how do we resolve the token conflicts?

Reply:
@damnever that is a good point that i should have touched more in the proposal. The idea is to have a similar approach to memberlist design.
The memberlist code resolves conflict is in the merge implementation of mergeable interface. The idea is to use a very similar code (maybe the same) of [resolveConflict](https://github.com/cortexproject/cortex/blob/48bc9001adf4dbe335b8fc14e6e7898bb94e8022/pkg/ring/model.go#L361) in the WatchKey for a new KV.

We still have the 'observe_period' on ring to verify any conflicts.
https://github.com/cortexproject/cortex/blob/48bc9001adf4dbe335b8fc14e6e7898bb94e8022/pkg/ring/lifecycler.go#L68